config.txt
#MAPNAME default.map
MAPNAME default.map

#AI LIST: AI <NAME> <STARTX> <STARTY> <UPIMAGE> <RIGHTIMAGE> <DOWNIMAGE> <LEFTIMAGE> <BULLETIMAGE>
AI SimpleAI 1 1
AI SimpleAI 15 4 images/tanks/Blue
AI SimpleAI 30 1 images/tanks/Red
AI SimpleAI 9 8 images/tanks/Green

AI SimpleAI 21 8 images/tanks/BlackWhite
AI SimpleAI 29 8 images/tanks/Coral
AI SimpleAI 4 11 images/tanks/Lime
AI SimpleAI 13 11 images/tanks/Purple

AI SimpleAI 23 11 images/tanks/SkyBlue
AI SimpleAI 15 9 images/tanks/Tiger


#FIELDIMAGE <MAP FILE>
FIELDIMAGE images/green.png

#OBSTACLE_IMAGE <IMAGE1> [<IMAGE2>]
OBSTACLE_IMAGE images/tree/treea.png images/rock/rocka.png
TREE_IMAGE images/tree/treea.png images/tree/treeb.png images/tree/treec.png
BUSH_IMAGE images/bush/busha.png images/bush/bushb.png images/bush/bushc.png
ROCK_IMAGE images/rock/rocka.png images/rock/rockb.png
WATER_IMAGE images/Water/waterTex.png images/Water/waterTexNS.png images/Water/waterTexES.png images/Water/waterTexSS.png images/Water/waterTexWS.png images/Water/waterTexNES.png images/Water/waterTexSES.png images/Water/waterTexSWS.png images/Water/waterTexNWS.png images/Water/waterTexEWS.png images/Water/waterTexNSS.png images/Water/waterTexNWES.png images/Water/waterTexNESS.png images/Water/waterTexSEWS.png images/Water/waterTexNWSS.png images/Water/waterTexPond.png

#Max Turns <200>
MAXTURNS 200

#TANK RULES
#STAT <AMMOUNT>
#VALID STATS: DAMAGE, HEALTH, RADAR, AP, SPECIAL, RANGE, AMMO
DAMAGE 2
HEALTH 3
RADAR 4
AP 2
SPECIAL 1
RANGE 4
AMMO 6

#AI IDLE SPEED: AI_SPEED <750>
AI_SPEED 750

#ANIMATION SPEED: ANIMATION_SPEED <20>
ANIMATION_SPEED 20

#BULLET SPEED: BULLET_SPEED <80>
BULLET_SPEED 80

#TANK SPEED: TANK_SPEED <400>
TANK_SPEED 400


src/config.txt
#AI LIST: AI <NAME> <STARTX> <STARTY>
AI SimpleAI 2 0 images/tank1_1.png images/tank1_2.png images/tank1_3.png images/tank1_4.png images/tank1_5.png images/tank1_6.png images/tank1_7.png images/tank1_8.png images/bullet1.png
AI SimpleAI 13 8 images/tank2_1.png images/tank2_2.png images/tank2_3.png images/tank2_4.png images/tank2_5.png images/tank2_6.png images/tank2_7.png images/tank2_8.png images/bullet2.png
#AI SimpleAI 14 2
#AI SimpleAI 1 8
#AI SimpleAI 13 8

#FIELD CONFIGURATION
WIDTH 15
HEIGHT 9

FIELDIMAGE images/green.png

OBSTACLE 1 0
OBSTACLE 5 7
OBSTACLE 12 2
OBSTACLE 3 7
OBSTACLE 9 6
OBSTACLE 13 5
OBSTACLE 2 5
OBSTACLE 10 1
OBSTACLE 8 3
OBSTACLE 2 2
OBSTACLE 5 5
OBSTACLE 14 7
OBSTACLE 4 2

OBSTACLE_IMAGE images/tree.png images/rock.png



#TANK RULES
DAMAGE 1
HEALTH 3
RANGE 2


changelog.md
# Slackers_Platform
Want instant communication with the Dev team? Join our
[Discord Channel](https://discord.gg/VN7ZuWR)!
## Release 7.00  ****  Major Changes to map system, AI's will need to be updated  ****
##  Changes
  - Smart pointers.  Most conventional points were switched to smart pointers for better memory management.
  - New command line options:
    -n  for no ui
    -q  for quiet text
    -t x  for x number of repeated games
    -h  added for command line option help and list of more commands
      * additional options can be checked in the help command or in the main.cpp
    - Changed -- for full word options and - for single option
    - Added support for chained single options.  IE -nq
  - No UI Mode.  The game can now be ran with no UI for extremely fast game times.  ~.05s average game time.
  - INI Parser. the config.txt is no more, now all options are in the config.ini
  - Map Loader. Maps are no longer input into the config, only the map name.  A separate utility will then load the
    map from the map directory.
  - Settings Class.  All settings were moved to their own class instead of spread across multiple sections.
  - Source file restructure.  Source files were moved into directories to better see what functionality they provide
  - Images.  Tanks now only use 3 images:  Base, Turret, and Bullet.  The base will rotate 360 degrees and the turret follows.
  - Added better text and descrptions to texture loading.
  - Rearranged texture loading to separate into user/system groups.
  - Improved AI loading process.
  - Functionalized texture loading.
  - Tank images moved to separate tank sub folder.
  - New tile object created for map.  The tile object contains the base for an object and then holds pointers for projectile or tank.
  - New Map.  The MapData object is now a single map made out of 2D vectors.  These vectors store Tile objects to give full information
        to users without multiple maps to check.
  - game init function massively reduced and improved.
  - Improved projectile turn speed.
  - Improved set special points functions and anti-cheating.
  - Added a game tracker that can output every move to a separate log file for tracking an entire game.
  - Added player count to loser section of results file.
  - Added option to specify name for results file.
  - Added command line option to specify different config file.
  - Cleaned up ActorInfo with reorganization and removed excess variables.
  - Improved create fog of war.
  - Rocks no longer respawn.
  - Cleaned up GameOver function.
  - Separated GameOver function to it's own file.
  - Functionalized some drawing functions for Tanks and Projectiles.  Added new draw function to Drawable.h/Drawable.cpp
  - Fixed/Improved SimpleAI after converting to use new Map system.
  - Other optimizations to help improve time.  ~.02 seconds is base time the additional .02-.03 is game time with 10 SimpleAI tanks.
## Release 6.05
##  Changes
  - Renamed README.md to changelog.md
  - Added -coverage option to command line.  This will set all speed values to 1 for a fast playing game.
  - runCoverage.sh will now actually perform coverage testing on the game files using tests in the  tests (src/functional) folder.
  - Changed testing binary names to reflect style of tests.
  - Additional make options added:  
    - make tests - creates the testing binaries in the main program folder
    - make clean-tests - cleans up testing
# FIXES
  - Fixed problem with CI script looking for old testing binary names.
  - Fixed problem with tanks folder not being created on a fresh pull and make.

## Release 6.03
##  Changes
  - Added -quiet command line option to remove a lot of output text
  - Added -demo command line option to increase game size for a 1080p screen
  - Fixed speed mode
    - Changing ANIMATION_SPEED, TANK_SPEED, PROJECTILE_SPEED to 0 should greatly decrease total game time.
## Release 6.02
## Changes
  - Changed output to results.txt file to improve human readablility.
    - The new format will list what turn the game ended on, the winner if there was one.  And then list each tank that participated in the match.  For each tank, their kills, hits, shots, AP, Max/Current Health, Max/Current Ammo, and final position is listed.

## Release 6.01
## Fixes
  - Fixed issue with Object map not always updating on a tank move command
  - Fixed issue with max_turns not being correctly enforced

## Release 6.00
## Fixes
  - Fixed projectiles turning into torpedos
  - Prevented rocks from regrowing on top of tanks
## Changes
  - Changed the default settings for the animation configuration.  The lower the number the faster the actor will move.  
## Balancing
- New Min and Max stat values:
             min       max      setAttribute
  - Damage    0   |     8     |   8
  - Health    0   |     8     |   8
  - AP        0   |     6     |   6
  - Radar     0   | Map Width | Map Width
  - Range     0   |    10     |   10
  - Ammo      0   |    10     |   -
  - Special   0   |    20     |   10
    - min - The minimum is the minimum value that can be set in the configuration file or assigned to your tank.  
      * Warning - If you do not add status to your tank when 0 is set, they will stay at 0, nothing will be automatically increased.
    - max - The maximum value that can be set in the config file
    - setAttribute - The max a stat can end at.  If this value is exceed the stat will be clamped down to the value in the chart.

## Release 5.01
## Fixes
- Non-Disappearing objects from obstacle and health map
  - Trees, Rocks, and (Future)Bushes will now disappear from the obstacle map when destroyed.
- Tree/Rock Regrowth (And future bushes)
  - These objects will now regrow after 4 turns back to their original health.  
  - They will also reappear back on the Obstacle and Health maps.
- Crates
  - Crates now properly get removed from the obstacle and health maps when they are destroyed.
  - They do NOT regrow.
- Health Adustments:

src/settings/Settings.h
/***************************************************************************//**
* @file settings.h
* @author Jonathan McKee
* @brief Contains all functions prototypes the settings class.
*******************************************************************************/
#ifndef __settings_h
#define __settings_h

#include <structures/attributes.h>
#include <string>

class GameField; //Forward declare it for pointer

/*!<Enum class for the gameMode */
enum gameMode {none, ai, sp, mp, quiet, coverage, tournament, noui};

/***************************************************************************//**
* @author Jonathan McKee
* @brief Class
*
* Settings is an interface for loading and hold all of the game settings
*******************************************************************************/
class Settings
{
public:
    //getters
    attributes getAttributes();
    int getIdleSpeed();
    int getAniFrames();
    int getTankSpeed();
    int getBulletSpeed();
    int getMaxTurns();
    gameMode getGameMode();
    std::string getResultsFile();
    std::string getConfigFile();
    std::string getMapName();

    //mode checkers
    bool checkCoverage();
    bool checkQuiet();
    bool showUI();
    bool checkTracking();

    //individual attribute getters
    int getAttrDamage();
    int getAttrHealth();
    int getAttrAP();
    int getAttrRadar();
    int getAttrAmmo();
    int getAttrRange();
    int getAttrSpecial();

    //setters
    void setAttributes(attributes attr);
    void setIdleSpeed(int is);
    void setAniFrames(int af);
    void setTankSpeed(int ts);
    void setBulletSpeed(int bs);
    void setMaxTurns(int mt);
    void setGameMode(gameMode gm);
    void setUI(bool set);
    void setCoverageMode(bool c);
    void setQuietMode(bool q);
    void setResultsFile(std::string r);
    void setConfigFile(std::string c);
    void setTrackingFile(std::string t);
    void setTrackingMode(bool t);
    void setMapName(std::string m);

    //individual attribute setters
    void setAttrDamage(int damage);
    void setAttrHealth(int health);
    void setAttrAP(int ap);
    void setAttrRadar(int radar, int width);
    void setAttrAmmo(int ammo);
    void setAttrRange(int range);
    void setAttrSpecial(int special);

    //Helper functions
    int checkSettingValue(int, int, int, std::string);
private:
    //default game settings
    int maxTurns = 1000;
    attributes baseAttributes;
    
    //Speed settings
    int idleSpeed = 750;
    int aniFrames = 20;
    int tankSpeed = 400;
    int bulletSpeed = 80;

    //pointers to components
    GameField *tankGame;   //Pointer to gamefield
    gameMode gMode = ai;  //default game mode

    //platform settings
    bool modeQuiet = false; //quiet mode for blocking text
    bool modeCoverage = false; //coverage mode for faster animations
    bool showui = true; //should we display the ui
    std::string resultsFile = "results.txt"; //results output file
    std::string configFile = "config.ini"; //Config ini file
    std::string mapName = "default.map"; //map file name
    bool tracking = false; //track all actions to file
};

#endif //__GAME_H
